src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs

```diff
@@ -515,6 +524,20 @@ private void CompileNamedType(NamedTypeSymbol containingType)
                             {
                                 hasStaticConstructor = true;
                             }
+
+                            if (_moduleBeingBuiltOpt is object && method is SourceMethodSymbolWithAttributes { MethodKind: MethodKind.Ordinary, IsModuleInitializer: true })
+                            {
+                                var rootModuleType = _moduleBeingBuiltOpt.RootModuleType;
+                                rootModuleType.AddModuleInitializerMethod(method);
```

Comment:
>method [](start = 74, length = 6)

Symbols should never be converted to CCI interfaces like that. They should be pushed through translation layer in PEModuleBuilder. I also think that there is no need to pass symbols around. We should be able to generate IL for the body and assign it. Is there a real reason for the rootModuleType to know what method will be called?

Reply:
> Symbols should never be converted to CCI interfaces like that. They should be pushed through translation layer in PEModuleBuilder.

Interesting, good to know. I might have further questions when I dig in to see what this means, since I've been leaning heavily on PrivateImplementationDetails to learn how codegen of global members is done.

> Is there a real reason for the rootModuleType to know what method will be called?

There needs to be a central location somewhere that coordinates the collection of module initializer methods as the source is visited concurrently. Then the central location needs to be frozen and the module initializer methods that were collected need to be sorted deterministically and emitted as calls inside `<Module>..cctor`. This doesn't necessarily have to happen inside RootModuleType, but it has to happen somewhere. I will need a suggestion of where.

Reply:
> There needs to be a central location somewhere that coordinates the collection of module initializer methods as the source is visited concurrently. Then the central location needs to be frozen and the module initializer methods that were collected need to be sorted deterministically and emitted as calls inside <Module>..cctor. This doesn't necessarily have to happen inside RootModuleType, but it has to happen somewhere. I will need a suggestion of where.

We usually collect things like that on the compilation level and often during binding rather than during emit phase, to avoid recalculating things over and over again. The first approach I would try is to accumulate valid candidates on a compilation in a thread-safe manner while we are processing well-known attributes. After declaration phase is done we would have all the candidates collected, guaranteed. We could "freeze" the collection at that point (end of declaration phase) as a mater of validation. During emit phase, without adding any symbol instances, I would find a good place to emit IL that calls all candidates in deterministic order, I am not sure if that would require synthesizing a bound tree or not. And would simply "communicate" to RootModuleType to inject .cctor with that IL.

---
In reply to: [409497028](https://github.com/dotnet/roslyn/pull/43281#discussion_r409497028) [](ancestors = 409497028)

Reply:
@AlekseyTs To collect initializer methods in a freezable collection on the compilation level, what do you think about decoding directly to a list in ModuleWellKnownAttributeData instead of MethodWellKnownAttributeData? It already has a 'freeze' concept. There could be an AddModuleInitializerMethod method and an ModuleInitializerMethods get-only property returning an immutable array.

Reply:
Awesome, MethodWellKnownAttributeData.AddNotNullMember is doing what I just described.

Reply:
> what do you think about decoding directly to a list in ModuleWellKnownAttributeData instead of MethodWellKnownAttributeData?

```ModuleWellKnownAttributeData ``` is about module level attributes and shouldn't be "polluted" with any other information.

Regarding ```MethodWellKnownAttributeData.AddNotNullMember```. saving the list on a method also doesn't feel like a good idea. We already should have several examples when we cache things on a compilation.

Reply:
Is SourceMethodSymbolWithAttributes.DecodeWellKnownAttributeAppliedToMethod the method that should be adding to the compilation, or should it still just cache the info in MethodWellKnownAttributeData? If the latter, what would the trigger be to read the bool on every method in the compilation and cache some of those methods in the compilation?

Reply:
> Is SourceMethodSymbolWithAttributes.DecodeWellKnownAttributeAppliedToMethod the method that should be adding to the compilation, or should it still just cache the info in ? If the latter, what would the trigger be to read the bool on every method in the compilation and cache some of those methods in the compilation?

There is no requirement to cache anything in MethodWellKnownAttributeData, unless we need a quick access to the information "coming" from an attribute. If we cache the method on a compilation when we process the attribute, do we really need anything from the method  in relation to the attribute later? It feels to me that we won't need anything.  So, performing validation and caching the method on a compilation should be sufficient for our scenario, I believe.

Reply:
Cool. I started going down this route. When I tried to write a test calling CSharpCompilation.GetModuleInitializerMethods, I realized I don't know the proper way to cause a pull so that all source methods in the compilation decode their well-known attributes. Is this something I should be doing (and do you know of an example I could look for) or should the semantics of GetModuleInitializerMethods() be such that it doesn't guarantee that all module initializer methods have been collected?

Reply:
> When I tried to write a test calling CSharpCompilation.GetModuleInitializerMethods, I realized I don't know the proper way to cause a pull so that all source methods in the compilation decode their well-known attributes. Is this something I should be doing (and do you know of an example I could look for) or should the semantics of GetModuleInitializerMethods() be such that it doesn't guarantee that all module initializer methods have been collected?

We control when this method is called. I assume we will be calling it during "compile" phase, that means that all declarations, including all attributes have been bound and analyzed, which means the list is fully populated. We can assert that inside the method by, for example, checking that ```_declarationDiagnosticsFrozen``` is true. In the code that adds candidates to the cache, we can assert that ```_declarationDiagnosticsFrozen``` is false. Alternatively you can add a dedicated flag, but that would probably be unnecessary, because you would have to set it to true at the same time ```_declarationDiagnosticsFrozen``` is set to true.

Reply:
> And would simply "communicate" to RootModuleType to inject .cctor with that IL.

Does this mean I should keep my [original SynthesizedRootModuleTypeStaticConstructor class](https://github.com/dotnet/roslyn/blob/4e4889adef072f44f0ed97ce096ddc55a629f931/src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedRootModuleTypeStaticConstructor.cs#L16)?

Reply:
> Does this mean I should keep my [original SynthesizedRootModuleTypeStaticConstructor class](https://github.com/dotnet/roslyn/blob/4e4889adef072f44f0ed97ce096ddc55a629f931/src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedRootModuleTypeStaticConstructor.cs#L16) and return it from RootModuleType.GetMethods?

It should be sufficient to implement Cci.IMethodDefinition, see for example Microsoft.CodeAnalysis.Emit.NoPia.VtblGap or CommonEmbeddedMethod. That new class should be "next to" RootModuleType and RootModuleType should return its instance from GetMethods.

Reply:
To obtain an IMethodBody to return from the new implementation of IMethodDefinition, should I find a way to call Microsoft.CodeAnalysis.CSharp.MethodCompiler.GenerateMethodBody or bypass it?

Reply:
Second question: the reason I started down the route of freezing things on RootModuleType was because RootModuleType and RootModuleStaticConstructor need to mutually reference each other via GetMethods and ContainingTypeDefinition. Unless I can safely return null from ContainingTypeDefinition?

If the mutual referencing has to stay, would I be correct that RootModuleType would need a an AddStaticConstructor method and a Freeze method?

Reply:
> To obtain an IMethodBody to return from the new implementation of IMethodDefinition, should I find a way to call Microsoft.CodeAnalysis.CSharp.MethodCompiler.GenerateMethodBody or bypass it? 

It looks like ```MethodCompiler.GenerateMethodBody``` has a lot of dependencies that we really do not have. I would prefer if we wouldn't try to "fake" all of them just so that method can be called. Either, we can go through the trouble of making all that input optional for the ```GenerateMethodBody``` and make sure it doesn't "die" (feels unnecessary complicated and risky), or we could use ```ILBuilder``` directly to emit a sequence of call instructions how ```CodeGen.CodeGenerator``` does that (we don't have to pass args or deal with return values, so should be relatively simple to do).    

> Calling `new MethodBody` directly does not look straightforward.

Again, I don't think we have to create an instance of ```MethodBody```, and we probably don't want to do that due to extra dependencies it has. It should be enough to have custom implementation of ```Cci.IMethodBody``` (see ```private sealed class EmptyBody : Cci.IMethodBody```, for example). The class for constructor can even implement ```IMethodDefinition``` and ```IMethodBody``` at the same time.

> Second question: the reason I started down the route of freezing things on RootModuleType was because RootModuleType and RootModuleStaticConstructor need to mutually reference each other via GetMethods and ContainingTypeDefinition. Unless I can safely return null from ContainingTypeDefinition?

> If the mutual referencing has to stay, would I be correct that RootModuleType would need a an AddStaticConstructor method and a Freeze method?

I would think that the ```RootModuleType``` would need some method to accept an IL blob for the constructor, which would trigger creation of ```IMethodDefinition``` instance for the constructor, and an allocation and caching of a non-empty collection of methods that ```GetMethods``` would return. That method should only ever be called once during lifetime of the ```RootModuleType``` instance. So, we should be able to detect a violation by simply checking if we cached something already. ```GetMethods``` would have to cache an empty collection if the cache is null. It doesn't look like we would need any explicit freezing mechanism.

Reply:
Implementing IMethodBody and IMethodDefinition leads to a dilemma because both interfaces inherit `IReference.Dispatch(MetadataVisitor visitor)` and I'm not sure if I can arbitrarily choose `visitor.Visit((IMethodDefinition)this);` versus `visitor.Visit((IMethodBody)this);` or simply throw.

Reply:
> Implementing IMethodBody and IMethodDefinition in the same class leads to a dilemma because both interfaces inherit `IReference.Dispatch(MetadataVisitor visitor)` and I'm not sure if I can arbitrarily choose `visitor.Visit((IMethodDefinition)this);` versus `visitor.Visit((IMethodBody)this);` or simply throw. Or maybe call both?

It doesn't look like ```IMethodBody``` inherits from ```IReference```
```
    internal interface IMethodBody
    {
```

What am I missing?

Reply:
Sorry, bad assumption because the visitor has an overload for IMethodBody.